rust/arrow/src/ffi.rs

```diff
@@ -193,6 +206,20 @@ fn to_datatype(format: &str) -> Result<DataType> {
         "ttm" => DataType::Time32(TimeUnit::Millisecond),
         "ttu" => DataType::Time64(TimeUnit::Microsecond),
         "ttn" => DataType::Time64(TimeUnit::Nanosecond),
+
+        // Note: The datatype null will only be created when called from ArrowArray::buffer_len
+        // at that point the child data is not yet known, but it is also not required to determine
+        // the buffer length of the list arrays.
+        "+l" => DataType::List(Box::new(Field::new(
+            "item",
+            child_type.unwrap_or(DataType::Null),
+            true,
+        ))),
+        "+L" => DataType::LargeList(Box::new(Field::new(
+            "item",
+            child_type.unwrap_or(DataType::Null),
+            true,
+        ))),
```

Comment:
@ritchie46 IMO the issue may be here: we need to re-build the entire field (name and nullability) from the schema, not just the datatype. cc @pitrou .

Reply:
Yes, and I also found that we need to send that to C++ as well. Currently we set the [flags](https://arrow.apache.org/docs/format/CDataInterface.html#c.ArrowSchema.flags) to 0 (omitting info). 

In the latest push I set it to `2` (nullable), and that fixed it. I will see if I can determine that information from the existing data.

Reply:
@jorgecarleitao the field is now built from the schema and we build the schema depending on the field!

This solved all the earlier issues. :)